src/Lecture2.hs

```diff
@@ -36,260 +42,145 @@ module Lecture2
     , Expr (..)
     , Variables
     , EvalError (..)
+    , State
     , eval
+    , combine
     , constantFolding
+    , constrExpr
     ) where
 
--- VVV If you need to import libraries, do it after this line ... VVV
-
--- ^^^ and before this line. Otherwise the test suite might fail  ^^^
+import Data.Char
+import Data.List
+import Data.Maybe
+import Data.Either
 
-{- | Implement a function that finds a product of all the numbers in
-the list. But implement a lazier version of this function: if you see
-zero, you can stop calculating product and return 0 immediately.
-
->>> lazyProduct [4, 3, 7]
-84
--}
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
+lazyProduct [] = 1
+lazyProduct (0 : _) = 0
+lazyProduct (x : xs) = x * lazyProduct xs
 
-{- | Implement a function that duplicates every element in the list.
-
->>> duplicate [3, 1, 2]
-[3,3,1,1,2,2]
->>> duplicate "cab"
-"ccaabb"
--}
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
-
-{- | Implement function that takes index and a list and removes the
-element at the given position. Additionally, this function should also
-return the removed element.
-
->>> removeAt 0 [1 .. 5]
-(Just 1,[2,3,4,5])
-
->>> removeAt 10 [1 .. 5]
-(Nothing,[1,2,3,4,5])
--}
-removeAt = error "TODO"
-
-{- | Write a function that takes a list of lists and returns only
-lists of even lengths.
-
->>> evenLists [[1,2,3], [3,1,2,7], [], [5, 7, 2]]
-[[3,1,2,7],[]]
-
-♫ NOTE: Use eta-reduction and function composition (the dot (.) operator)
-  in this function.
--}
-evenLists = error "TODO"
-
-{- | The @dropSpaces@ function takes a string containing a single word
-or number surrounded by spaces and removes all leading and trailing
-spaces.
+duplicate [] = []
+duplicate (x : xs) = x : x : duplicate xs
 
->>> dropSpaces "   hello  "
-"hello"
->>> dropSpaces "-200            "
-"-200"
+removeAt::Int -> [a] -> (Maybe a, [a])
+removeAt _ [] = (Nothing,[])
+removeAt 0 (x : xs) = (Just x, xs)
+removeAt n (x : xs)
+    | n > 0 = let (v, rem) = removeAt (n -1) xs in (v, x : rem)
+	| otherwise = (Nothing,(x : xs))
 
-♫ NOTE: As in the previous task, use eta-reduction and function
-  composition (the dot (.) operator) in this function.
-
-🕯 HINT: look into Data.Char and Prelude modules for functions you may use.
--}
-dropSpaces = error "TODO"
+evenLists::[[a]] -> [[a]]
+evenLists [] = []
+evenLists [[]] = [[]]
+evenLists ((x:xs):tx) = [(x:xs) | (x:xs) <- ((x:xs):tx),  (even.length) (x:xs)]
```

Comment:
You can implement this function simpler by using a standard function `filter` instead and passing the composition of `even` and `length` to it 👍🏻